src/java/org/apache/cassandra/cql3/functions/TimeFcts.java

```diff
@@ -101,94 +136,154 @@ public ByteBuffer execute(ProtocolVersion protocolVersion, List<ByteBuffer> para
+        }
+    }
+
+    /**
+     * Creates a function that converts a value of the specified type into a {@code DATE}.
+     *
+     * @param type the temporal type
+     * @return a function that convert a value of the specified type into a <code>DATE</code>.
+     */
+    public static NativeScalarFunction toDate(TemporalType<?> type)
+    {
+        return new ToDateFunction(type, false);
+    }
+
+    private static class ToDateFunction extends NativeScalarFunction
+    {
+        private final TemporalType<?> type;
+
+        public ToDateFunction(TemporalType<?> type, boolean useLegacyName)
+        {
+            super(useLegacyName ? "to_date" : "todate", SimpleDateType.instance, type);
+            this.type = type;
+        }
+
+        @Override
+        public ByteBuffer execute(ProtocolVersion protocolVersion, List<ByteBuffer> parameters)
+        {
+            ByteBuffer bb = parameters.get(0);
+            if (bb == null || !bb.hasRemaining())
+                return null;
+
+            long millis = type.toTimeInMillis(bb);
+            return SimpleDateType.instance.fromTimeInMillis(millis);
+        }
+
+        @Override
+        public boolean isMonotonic()
+        {
+            return true;
+        }
+
+        @Override
+        public NativeFunction withLegacyName()
+        {
+            return new ToDateFunction(type, true);
+        }
+    }
 
     /**
-     * Creates a function that convert a value of the specified type into an UNIX timestamp.
+     * Creates a function that converts a value of the specified type into a {@code TIMESTAMP}.
+     *
      * @param type the temporal type
-     * @return a function that convert a value of the specified type into an UNIX timestamp.
+     * @return a function that convert a value of the specified type into a {@code TIMESTAMP}.
      */
-    public static final NativeScalarFunction toUnixTimestamp(final TemporalType<?> type)
+    public static NativeScalarFunction toTimestamp(TemporalType<?> type)
     {
-        return new NativeScalarFunction("tounixtimestamp", LongType.instance, type)
+        return new ToTimestampFunction(type, false);
+    }
+
+    private static class ToTimestampFunction extends NativeScalarFunction
+    {
+        private final TemporalType<?> type;
+
+        public ToTimestampFunction(TemporalType<?> type, boolean useLegacyName)
         {
-            public ByteBuffer execute(ProtocolVersion protocolVersion, List<ByteBuffer> parameters)
-            {
-                ByteBuffer bb = parameters.get(0);
-                if (bb == null || !bb.hasRemaining())
-                    return null;
-
-                return ByteBufferUtil.bytes(type.toTimeInMillis(bb));
-            }
-
-            @Override
-            public boolean isMonotonic()
-            {
-                return true;
-            }
-        };
+            super(useLegacyName ? "to_timestamp" : "totimestamp", TimestampType.instance, type);
```

Comment:
Shouldn't this be reversed?

Reply:
It should, good catch. I think I had a some point the inverse condition (`useNewName`) and forgot to update on refactor. Incidentally it wasn't breaking anything because in the end we add both names to the collection of native functions and we don't make any distinction between them. I have fixed it and added a new check to the existing test.